kotlin-analysis-api/src/main/kotlin/com/google/devtools/ksp/impl/symbol/kotlin/KSClassifierReferenceImpl.kt

```diff
@@ -1,37 +1,77 @@
 package com.google.devtools.ksp.impl.symbol.kotlin
 
 import com.google.devtools.ksp.IdKeyPair
+import com.google.devtools.ksp.IdKeyTriple
 import com.google.devtools.ksp.KSObjectCache
 import com.google.devtools.ksp.symbol.*
-import org.jetbrains.kotlin.analysis.api.components.buildClassType
-import org.jetbrains.kotlin.analysis.api.types.KtUsualClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassType
+import org.jetbrains.kotlin.analysis.api.types.KtClassTypeQualifier
+import org.jetbrains.kotlin.analysis.api.types.KtTypeParameterType
 
 class KSClassifierReferenceImpl private constructor(
-    internal val ktType: KtUsualClassType,
+    internal val ktType: KtClassType,
+    internal val index: Int,
     override val parent: KSTypeReference?
 ) : KSClassifierReference {
-    companion object : KSObjectCache<IdKeyPair<KtUsualClassType, KSTypeReference?>, KSClassifierReferenceImpl>() {
-        fun getCached(ktType: KtUsualClassType, parent: KSTypeReference?) =
-            cache.getOrPut(IdKeyPair(ktType, parent)) { KSClassifierReferenceImpl(ktType, parent) }
+    companion object : KSObjectCache<IdKeyTriple<KtClassType, Int, KSTypeReference?>, KSClassifierReferenceImpl>() {
+        fun getCached(ktType: KtClassType, parent: KSTypeReference?, index: Int) =
+            cache.getOrPut(IdKeyTriple(ktType, index, parent)) { KSClassifierReferenceImpl(ktType, index, parent) }
     }
+
+    private val classifierReference: KtClassTypeQualifier
+        get() = ktType.qualifiers[index]
+
     override val qualifier: KSClassifierReference? by lazy {
-        ktType.classId.outerClassId?.let {
-            analyze {
-                buildClassType(it)
-            }
-        }?.let { getCached(it as KtUsualClassType, parent) }
+        if (index == 0) {
+            null
+        } else {
+            getCached(ktType, parent, index - 1)
+        }
     }
 
     override fun referencedName(): String {
-        return ktType.classId.asFqNameString()
+        return classifierReference.name.asString()
     }
 
     override val typeArguments: List<KSTypeArgument> by lazy {
-        ktType.typeArguments().map { KSTypeArgumentImpl.getCached(it, this) }
+        classifierReference.typeArguments.map { KSTypeArgumentImpl.getCached(it, this) }
```

Comment:
I wonder if `KtNonErrorClassType.typeArguments` is just renamed to `KtNonErrorClassType.ownTypeArguments`?

This change seems like, picking the specific classifier reference at `index` and retrieving _that_ one's type arguments, instead of loading the type arguments of `ktType`.

Reply:
Yes, IIRC the change to rename `KtNonErrorClassType.typeArguments` to `KtNonErrorClassType.ownTypeArguments` was introduced together with the qualifier.

Reply:
So, my question was whether this is supposed to be `ktType.ownTypeArguments.map { ... }`, not `classifierReference.typeArguments.map { ... }`.

Reply:
No, it should be the latter because qualifier's argument only look at the arguments of its part. actually this `classifierReference` is not of a `KtType` class, but its own one.